FirebaseMLModelDownloader/Sources/TelemetryLogger.swift

```diff
@@ -0,0 +1,163 @@
+extension ModelDownloadLogEvent {
+  mutating func setEvent(status: DownloadStatus, errorCode: ErrorCode? = nil,
+                         roughDownloadDuration: UInt64? = nil, exactDownloadDuration: UInt64? = nil,
+                         downloadFailureStatus: Int64? = nil, modelOptions: ModelOptions) {
+    downloadStatus = status
+    if let code = errorCode {
+      self.errorCode = code
+    }
+    if let roughDuration = roughDownloadDuration {
+      roughDownloadDurationMs = roughDuration
+    }
+    if let exactDuration = exactDownloadDuration {
+      exactDownloadDurationMs = exactDuration
+    }
+    if let failureStatus = downloadFailureStatus {
+      self.downloadFailureStatus = failureStatus
+    }
+    options = modelOptions
+  }
+}
+
+/// Extension to build Firebase ML log event.
+extension FirebaseMlLogEvent {
+  mutating func setEvent(eventName: EventName, systemInfo: SystemInfo,
+                         modelDownloadLogEvent: ModelDownloadLogEvent) {
+    self.eventName = eventName
+    self.systemInfo = systemInfo
+    self.modelDownloadLogEvent = modelDownloadLogEvent
+  }
+}
+
+/// Data object for Firelog event.
+class FBMLDataObject: NSObject, GDTCOREventDataObject {
+  private let event: FirebaseMlLogEvent
+
+  init(event: FirebaseMlLogEvent) {
+    self.event = event
+  }
+
+  /// Encode Firelog event for transport.
+  func transportBytes() -> Data {
+    do {
+      // TODO: Should this be binary or json serialized?
+      let data = try event.serializedData()
+      print(try event.jsonString())
+      return data
+    } catch {
+      DeviceLogger.logEvent(
+        level: .debug,
+        category: .analytics,
+        message: "Unable to encode Firelog event.",
+        messageCode: .analyticsEventEncodeError
+      )
+      return Data()
+    }
+  }
+}
+
+/// Firelog logger.
+class TelemetryLogger {
+  private let mappingID = "1326"
+  let isStatsEnabled: Bool
+  let fllTransport: GDTCORTransport
+
+  private let apiKey: String?
+  private let projectID: String?
+
+  /// Init logger, could be nil if unable to get event transport.
+  init?(isStatsEnabled: Bool, apiKey: String?, projectID: String?) {
+    self.isStatsEnabled = isStatsEnabled
+    self.apiKey = apiKey
+    self.projectID = projectID
+    guard let fllTransport = GDTCORTransport(
+      mappingID: mappingID,
+      transformers: nil,
+      target: GDTCORTarget.FLL
+    ) else {
+      DeviceLogger.logEvent(
+        level: .debug,
+        category: .analytics,
+        message: "Unable to create telemetry logger.",
+        messageCode: .telemetryInitError
+      )
+      return nil
+    }
+    self.fllTransport = fllTransport
+  }
+
+  /// Log events to Firelog.
+  private func logModelEvent(event: FirebaseMlLogEvent) {
+    let eventForTransport: GDTCOREvent = fllTransport.eventForTransport()
+    eventForTransport.dataObject = FBMLDataObject(event: event)
+    eventForTransport.qosTier = .qoSFast
```

Comment:
I assume it is needed for debugging only. Could you please add a TODO to update QoS.

Reply:
Good catch! Fixed.